File: boj/TopologicalSorting/BOJ_2623.py
# 백준 2623 음악프로그램
# 21.03.16
import sys
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
rd = sys.stdin.readline
INF = int(1e9)

n, m = map(int, input().split())
indegree = [0] * (n + 1)
edges = [[] for _ in range(n + 1)]
for _ in range(m):
    temp = list(map(int, rd().split()))
    for i in range(2, len(temp)):
        edges[temp[i-1]].append(temp[i])
        indegree[temp[i]] += 1

ans = []
q = []
for i in range(1, n + 1):
    if indegree[i] == 0:
        q.append(i)
        ans.append(i)

while q:
    now = q.pop(0)

    for i in edges[now]:
        indegree[i] -= 1
        if indegree[i] == 0:
            q.append(i)
            ans.append(i)

if len(ans) < n: print(0)
else:
    for i in ans: print(i)

File: boj/TopologicalSorting/BOJ_2623.swift
// 백준 2623 음악프로그램
// 21.03.23

import Foundation

let nm = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = nm[0]
let m = nm[1]

var indegree = [Int](repeating: 0, count: n + 1)
var ans = [Int]()
var graph = [[Int]](repeating: [Int](), count: n + 1)

for _ in 0..<m {
  let arr = readLine()!.split(separator: " ").map { Int(String($0))! }
  
  for i in stride(from: 2, through: arr[0], by: 1) {
    graph[arr[i-1]].append(arr[i])
    indegree[arr[i]] += 1
  }
}

var q = [Int]()

for i in stride(from: 1, through: n, by: 1) {
  if indegree[i] == 0 {
    q.append(i)
    ans.append(i)
  }
}

var cnt = 0
while !q.isEmpty {
  let now = q.removeLast()
  cnt += 1
  
  for i in graph[now] {
    indegree[i] -= 1
    if indegree[i] == 0 {
      ans.append(i)
      q.append(i)
    }
  }
}

if cnt < n { print(0) }
else { for i in ans { print(i) }}
